go/enclave/services/sequencer.go

```diff
@@ -0,0 +1,324 @@
+package services
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/rand"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	gethlog "github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+	"github.com/obscuronet/go-obscuro/go/common"
+	"github.com/obscuronet/go-obscuro/go/common/errutil"
+	"github.com/obscuronet/go-obscuro/go/common/log"
+	"github.com/obscuronet/go-obscuro/go/enclave/components"
+	"github.com/obscuronet/go-obscuro/go/enclave/core"
+	"github.com/obscuronet/go-obscuro/go/enclave/crypto"
+	"github.com/obscuronet/go-obscuro/go/enclave/db"
+	"github.com/obscuronet/go-obscuro/go/enclave/mempool"
+)
+
+type sequencer struct {
+	blockConsumer  components.L1BlockProcessor
+	batchProducer  components.BatchProducer
+	batchRegistry  components.BatchRegistry
+	rollupProducer components.RollupProducer
+	rollupConsumer components.RollupConsumer
+
+	logger gethlog.Logger
+
+	hostID            gethcommon.Address
+	chainConfig       *params.ChainConfig
+	enclavePrivateKey *ecdsa.PrivateKey // this is a key known only to the current enclave, and the public key was shared with everyone during attestation
+	mempool           mempool.Manager
+	storage           db.Storage
+	encryption        crypto.TransactionBlobCrypto
+
+	// This is used to coordinate creating
+	// new batches and creating fork batches.
+	batchProductionMutex sync.Mutex
+}
+
+func NewSequencer(
+	consumer components.L1BlockProcessor,
+	producer components.BatchProducer,
+	registry components.BatchRegistry,
+	rollupProducer components.RollupProducer,
+	rollupConsumer components.RollupConsumer,
+
+	logger gethlog.Logger,
+
+	hostID gethcommon.Address,
+	chainConfig *params.ChainConfig,
+	enclavePrivateKey *ecdsa.PrivateKey, // this is a key known only to the current enclave, and the public key was shared with everyone during attestation
+	mempool mempool.Manager,
+	storage db.Storage,
+	encryption crypto.TransactionBlobCrypto,
+) Sequencer {
+	return &sequencer{
+		blockConsumer:     consumer,
+		batchProducer:     producer,
+		batchRegistry:     registry,
+		rollupProducer:    rollupProducer,
+		rollupConsumer:    rollupConsumer,
+		logger:            logger,
+		hostID:            hostID,
+		chainConfig:       chainConfig,
+		enclavePrivateKey: enclavePrivateKey,
+		mempool:           mempool,
+		storage:           storage,
+		encryption:        encryption,
+	}
+}
+
+func (s *sequencer) CreateBatch() error {
```

Comment:
Would be cool to have unit tests around this where we mock bad data, forks or other weird scenarios and see the error / expected batches come out.